src/Data/Integer/ModularArithmetic.agda

```diff
@@ -0,0 +1,96 @@
+ℤ/mℤ  = quotientRing
+
+-- todo:
+-- * chinese remainder theorem specialized to integers
+-- * finiteness: for non-zero
+
+module _ .{{_ : NonZero m}} where
+
+  from-% : ∀ {x y} → x % m ≡ y % m → x ≋ y
+  from-% {x} {y} x%m≡y%m = x / m - y / m by begin
+    x - y                                                   ≡⟨ ≡.cong₂ _-_ (a≡a%n+[a/n]*n x m) (a≡a%n+[a/n]*n y m) ⟩
+    (+ (x % m) + (x / m) * m) - (+ (y % m) + (y / m) * m)   ≡⟨ ≡.cong (λ z → (+ (x % m) + (x / m) * m) + z) (neg-distrib-+ (+ (y % m)) ((y / m) * m)) ⟩
+    (+ (x % m) + (x / m) * m) + (- + (y % m) - (y / m) * m) ≡⟨ +-assoc (+ (x % m) + (x / m) * m) (- + (y % m)) (- ((y / m) * m)) ⟨
+    ((+ (x % m) + (x / m) * m) - + (y % m)) - (y / m) * m   ≡⟨ ≡.cong (λ z → z - (y / m) * m) (+-assoc (+ (x % m)) ((x / m) * m) (- + (y % m))) ⟩
+    (+ (x % m) + ((x / m) * m - + (y % m))) - (y / m) * m   ≡⟨ ≡.cong (λ z → (+ (x % m) + z) - (y / m) * m) (+-comm ((x / m) * m) (- + (y % m))) ⟩
+    (+ (x % m) + (- + (y % m) + (x / m) * m)) - (y / m) * m ≡⟨ ≡.cong (λ z → z - (y / m) * m) (+-assoc (+ (x % m)) (- + (y % m)) ((x / m) * m)) ⟨
+    ((+ (x % m) - + (y % m)) + (x / m) * m) - (y / m) * m   ≡⟨ +-assoc (+ (x % m) - + (y % m)) ((x / m) * m) (- ((y / m) * m)) ⟩
+    (+ (x % m) - + (y % m)) + ((x / m) * m - (y / m) * m)   ≡⟨ ≡.cong₂ (λ a b → (+ a - + (y % m)) + ((x / m) * m + b)) x%m≡y%m (neg-distribˡ-* (y / m) m) ⟩
+    (+ (y % m) - + (y % m)) + ((x / m) * m + - (y / m) * m) ≡⟨ ≡.cong₂ _+_ (+-inverseʳ (+ (y % m))) (≡.sym (*-distribʳ-+ m (x / m) (- (y / m)))) ⟩
+    0ℤ + (x / m - y / m) * m                                ≡⟨ +-identityˡ ((x / m - y / m) * m) ⟩
+    (x / m - y / m) * m                                     ∎
+    where open ≡-Reasoning
+
+  to-% : ∀ {x y} → x ≋ y → x % m ≡ y % m
+  to-% {x} {y} (k by x-y≡km) = {!!}
+    where
+      open ≡-Reasoning
+      lemma : x % m ⊖ y % m ≡ (k - (x / m) + (y / m)) * m
+      lemma = begin
+        x % m ⊖ y % m               ≡⟨ m-n≡m⊖n (x % m) (y % m) ⟨
+        + (x % m) - + (y % m)       ≡⟨ {!!} ⟩
+        (k - (x / m) + (y / m)) * m ∎
+
+      bound : ∣ x % m ⊖ y % m ∣ ℕ.< ∣ m ∣
+      bound = {!!}
+
+  _≋?′_ : Decidable _≋_
+  x ≋?′ y = map′ from-% to-% (x % m ℕ.≟ y % m)
+
+  ℤ/mℤ-finite : Equivalence (Ring.setoid ℤ/mℤ) (≡.setoid (Fin ∣ m ∣))
+  ℤ/mℤ-finite = record
+    { to = λ n → Fin.fromℕ< (n%d<d n m)
+    ; from = λ i → + Fin.toℕ i
+    ; to-cong = λ x≋y → Fin.fromℕ<-cong _ _ (to-% x≋y) _ _
+    ; from-cong = λ i≡j → Ring.reflexive ℤ/mℤ (≡.cong (+_ ∘ Fin.toℕ) i≡j)
+    }
+    where
+      open ≡.≡-Reasoning
+
+      -- x ≡ x % m + (x / m) * m
+      -- y ≡ y % m + (y / m) * m
+      -- x - y ≡ (x / m - y / m) * m + x % m - y % m
+      -- x - y ≡ k * m
+      -- something about x % m being smaller than m and at least 0?
+      -- so k ≡ x / m - y / m
+      -- and so x % m - y % m
+      -- ∴ x % m ≡ y % m
+
+_≋?_ : Decidable _≋_
+-- need to check whether m is 0
+_≋?_ with ℕ.nonZero? ∣ m ∣
+... | yes p = _≋?′_ {{p}}
+... | no ¬p = {!!}
```

Comment:
My first thought on reading this code is that `.{{_ : NonZero m}}` should in fact also be part of the `module` parametrisation, and thus
* the anonymous `module _ .{{_ : NonZero m}} where` can be removed
* `_≋?_` just *is* `_≋?′_`

Reply:
I'm not sure why we need to treat the m = 1 case specially. It's not a particularly interesting case, but it should work the same as any other nonzero modulus

Reply:
Yes, that was my plan